src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,453 @@
+  }
+  for (const double* block : point_blocks_) {
+    eval_options.parameter_blocks.push_back(const_cast<double*>(block));
+  }
+  ceres::CRSMatrix J_full_crs;
+  problem_->Evaluate(eval_options, nullptr, nullptr, nullptr, &J_full_crs);
+  int num_residuals = J_full_crs.num_rows;
+  int num_params = J_full_crs.num_cols;
+  Eigen::Map<Eigen::SparseMatrix<double, Eigen::RowMajor>> J_full(
+      J_full_crs.num_rows,
+      J_full_crs.num_cols,
+      J_full_crs.values.size(),
+      J_full_crs.rows.data(),
+      J_full_crs.cols.data(),
+      J_full_crs.values.data());
+
+  // Step 3: Schur elimination on points
+  LOG(INFO) << "Schur elimination on points";
+  Eigen::SparseMatrix<double> J_c =
+      J_full.block(0, 0, num_residuals, num_params - num_params_points_);
+  Eigen::SparseMatrix<double> J_p = J_full.block(
+      0, num_params - num_params_points_, num_residuals, num_params_points_);
+  Eigen::SparseMatrix<double> H_cc = J_c.transpose() * J_c;
+  Eigen::SparseMatrix<double> H_cp = J_c.transpose() * J_p;
+  Eigen::SparseMatrix<double> H_pc = H_cp.transpose();
+  Eigen::SparseMatrix<double> H_pp = J_p.transpose() * J_p;
+  // in-place computation of H_pp_inv
+  Eigen::SparseMatrix<double>& H_pp_inv = H_pp;  // reference
+  int counter_p = 0;
+  for (const double* block : point_blocks_) {
+    int num_params_point = ParameterBlockTangentSize(problem_, block);
+    Eigen::SparseMatrix<double> subMatrix_sparse =
+        H_pp.block(counter_p, counter_p, num_params_point, num_params_point);
+    Eigen::MatrixXd subMatrix = subMatrix_sparse;
+    subMatrix +=
+        lambda_ * Eigen::MatrixXd::Identity(subMatrix.rows(), subMatrix.cols());
+    Eigen::MatrixXd subMatrix_inv = subMatrix.inverse();
+    // update matrix
+    for (int i = 0; i < num_params_point; ++i) {
+      for (int j = 0; j < num_params_point; ++j) {
+        H_pp_inv.coeffRef(counter_p + i, counter_p + j) = subMatrix_inv(i, j);
+      }
+    }
+    counter_p += num_params_point;
+  }
+  H_pp_inv.makeCompressed();
+  S_matrix_ = H_cc - H_cp * H_pp_inv * H_pc;
+}
+
+bool BundleAdjustmentCovarianceEstimator::HasValidSchurComplement() const {
+  return S_matrix_.size() != 0;
+}
+
+bool BundleAdjustmentCovarianceEstimator::ComputeFull() {
+  if (!HasValidSchurComplement()) {
+    ComputeSchurComplement();
+  }
+  Eigen::MatrixXd S_dense = S_matrix_;
+  LOG(INFO) << StringPrintf(
+      "Inverting Schur complement for all variables except for 3D points (n = "
+      "%d)",
+      num_params_poses_ + num_params_other_variables_);
+  Eigen::FullPivLU<Eigen::MatrixXd> luOfS_dense(S_dense);
+  if (luOfS_dense.rank() < S_dense.rows()) {
+    LOG(INFO) << StringPrintf(
+        "Unable to compute covariance. The Schur complement on all variables "
+        "(except for 3D points) is rank deficient. Number of columns: %d, "
+        "rank: %d.",
+        S_dense.rows(),
+        luOfS_dense.rank());
+    return false;
+  }
+  cov_variables_ = S_dense.inverse();
```

Comment:
Can we share this block of code between `Compute` and `ComputeFull`?

Reply:
The error logs are slightly different. There is no other cases for using this inverse so I would vote for keeping it as in the current form.

Reply:
You could handle the logging of the error in the respective parent functions?

Reply:
We print the rank here in the logs, and this will require either exposing some intermediate variables or making the rank output in the inverse function. I feel that this abstraction is a little overcomplication. 

Also, here https://github.com/colmap/colmap/pull/2610/commits/9b3ea5f6618d77fdaf47ae57f54de86d13189652 I have updated the matrix inverse for ``ComputeFull`` to use sparse cholesky while use dense LU for handling only pose blocks at ``Compute``. So there is no point to abstract this part anymore.

Reply:
Nice, thanks!